src/hotspot/share/oops/klass.hpp

```diff
@@ -523,7 +523,7 @@ class Klass : public Metadata {
            "sanity. l2esz: 0x%x for lh: 0x%x", (uint)l2esz, (uint)lh);
     return l2esz;
   }
-  static jint array_layout_helper(jint tag, int hsize, BasicType etype, int log2_esize) {
+  static jint array_layout_helper(uint32_t tag, int hsize, BasicType etype, int log2_esize) {
```

Comment:
```suggestion
  static jint array_layout_helper(jint tag, int hsize, BasicType etype, int log2_esize) {
    tag &= right_n_bits(_lh_array_tag_bits);
```
I'd like to propose a simpler fix that doesn't involve changing to unsigned.  In my experiments this satisfies UBSAN while still generating the same optimized code in product builds.